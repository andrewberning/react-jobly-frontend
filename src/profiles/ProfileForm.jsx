import { useState, useContext, useEffect } from "react"
import Alert from "../components/Alert"
import JoblyApi from "../api/api"
import UserContext from "../auth/UserContext"

export default function ProfileForm() {
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const [formData, setFormData] = useState({
    firstName: "",
    lastName: "",
    email: "",
    username: "",
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  const [saveConfirmed, setSaveConfirmed] = useState(false);

  useEffect(() => {
    if (currentUser) {
      setFormData({
        firstName: currentUser.first_name || "",
        lastName: currentUser.last_name || "",
        email: currentUser.email || "",
        username: currentUser.username || "",
        password: "",
      });
    }
  }, [currentUser]);

  async function handleSubmit(evt) {
    evt.preventDefault();

    let profileData = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: currentUser.email,
      password: formData.password,
    };

    let username = formData.username;
    let updatedUser;

    try {
      updatedUser = await JoblyApi.saveProfile(username, profileData);
    } catch(error) {
      setFormErrors([error.message]);
      return;
    }

    setFormData(f => ({ ...f, password: "" }));
    setFormErrors([]);
    setSaveConfirmed(true);

    setCurrentUser(updatedUser);
  }

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(f => ({
      ...f,
      [name]: value,
    }));
    setFormErrors([]);
  }

  return (
    <div className="ProfileForm col-md-6">
      <h3>Profile</h3>
      <div className="card">
        <div className="card-body">
          <form onSubmit={handleSubmit}>
            <div className="form-group">
              <label htmlFor="username">Username</label>
              <p className="form-control-plaintext" id="username">{formData.username}</p>
            </div>
            <div className="form-group">
              <label htmlFor="firstName">First Name</label>
              <input 
                id="firstName"
                name="firstName"
                className="form-control"
                value={formData.firstName}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="lastName">Last Name</label>
              <input 
                id="lastName"
                name="lastName"
                className="form-control"
                value={formData.lastName}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
              <label>Email</label>
              <input 
                name="email"
                className="form-control"
                value={formData.email}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
                <label htmlFor="password">Confirm password to make changes:</label>
                <input
                    type="password"
                    id="password"
                    name="password"
                    className="form-control"
                    value={formData.password}
                    onChange={handleChange}
                />
            </div>

            {formErrors.length
                  ? <Alert type="danger" messages={formErrors} />
                  : null}

              {saveConfirmed
                  ?
                  <Alert type="success" messages={["Updated successfully."]} />
                  : null}


              <button
                  className="btn btn-primary btn-block mt-4"
                  onClick={handleSubmit}
              >
                Save Changes
              </button>
          </form>
        </div>
      </div>
    </div>
  )
}