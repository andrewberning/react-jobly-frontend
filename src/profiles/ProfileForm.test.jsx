import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import ProfileForm from './ProfileForm';
import { UserProvider } from "../testUtils";
import JoblyApi from '../api/api';

// Mock the Alert component
vi.mock('../components/Alert', () => ({
  __esModule: true,
  default: ({ type, messages }) => (
    <div data-testid={`alert-${type}`}>{messages.join(' ')}</div>
  ),
}));

// Mock the JoblyApi
vi.mock('../api/api', () => ({
  __esModule: true,
  default: {
    saveProfile: vi.fn(),
  },
}));

describe('ProfileForm', () => {

  const setCurrentUser = vi.fn();

  const renderProfileForm = (currentUser) =>
    render(
      <UserProvider currentUser={currentUser} setCurrentUser={setCurrentUser}>
        <ProfileForm />
      </UserProvider>
    );

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <UserProvider>
        <ProfileForm />
      </UserProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  })

  it('renders the form with user data', () => {
    renderProfileForm();
    
    expect(screen.getByDisplayValue('testfirst')).toBeInTheDocument();
    expect(screen.getByDisplayValue('testlast')).toBeInTheDocument();
    expect(screen.getByDisplayValue('test@test.net')).toBeInTheDocument();
    expect(screen.getByText('testuser')).toBeInTheDocument();
  });

  it('handles input changes', () => {
    renderProfileForm();
    
    fireEvent.change(screen.getByLabelText(/First Name/i), {
      target: { value: 'Jane' },
    });
    fireEvent.change(screen.getByLabelText(/Last Name/i), {
      target: { value: 'Smith' },
    });

    expect(screen.getByDisplayValue('Jane')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Smith')).toBeInTheDocument();
  });

  it('submits the form successfully', async () => {
    JoblyApi.saveProfile.mockResolvedValueOnce({
      username: 'testuser',
      firstName: 'Jane',
      lastName: 'Smith',
      email: 'test@test.net',
      photo_url: null,
    });

    renderProfileForm();

    fireEvent.change(screen.getByLabelText(/First Name/i), {
      target: { value: 'Jane' },
    });
    fireEvent.change(screen.getByLabelText(/Last Name/i), {
      target: { value: 'Smith' },
    });
    fireEvent.change(screen.getByLabelText(/Confirm password/i), {
      target: { value: 'password123' },
    });

    fireEvent.click(screen.getByText(/Save Changes/i));

    await waitFor(() => {
      expect(JoblyApi.saveProfile).toHaveBeenCalledWith('testuser', {
        firstName: 'Jane',
        lastName: 'Smith',
        email: 'test@test.net',
        password: 'password123',
      });
      expect(screen.getByTestId('alert-success')).toBeInTheDocument();
    });

    expect(setCurrentUser).toHaveBeenCalledWith({
      username: 'testuser',
      firstName: 'Jane',
      lastName: 'Smith',
      email: 'test@test.net',
      photo_url: null,
    });
  });

  it('handles form submission errors', async () => {
    JoblyApi.saveProfile.mockRejectedValueOnce(new Error('Error updating profile'));

    renderProfileForm();

    fireEvent.change(screen.getByLabelText(/First Name/i), {
      target: { value: 'Jane' },
    });
    fireEvent.change(screen.getByLabelText(/Last Name/i), {
      target: { value: 'Smith' },
    });
    fireEvent.change(screen.getByLabelText(/Confirm password/i), {
      target: { value: 'password123' },
    });

    fireEvent.click(screen.getByText(/Save Changes/i));

    await waitFor(() => {
      expect(JoblyApi.saveProfile).toHaveBeenCalledWith('testuser', {
        firstName: 'Jane',
        lastName: 'Smith',
        email: 'test@test.net',
        password: 'password123',
      });
      expect(screen.getByTestId('alert-danger')).toBeInTheDocument();
    });

    expect(setCurrentUser).not.toHaveBeenCalled();
  });
});
