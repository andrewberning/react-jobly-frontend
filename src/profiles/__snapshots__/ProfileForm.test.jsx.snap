// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ProfileForm col-md-6"
  >
    <h3>
      Profile
    </h3>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              for="username"
            >
              Username
            </label>
            <p
              class="form-control-plaintext"
              id="username"
            >
              testuser
            </p>
          </div>
          <div
            class="form-group"
          >
            <label
              for="firstName"
            >
              First Name
            </label>
            <input
              class="form-control"
              id="firstName"
              name="firstName"
              value="testfirst"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="lastName"
            >
              Last Name
            </label>
            <input
              class="form-control"
              id="lastName"
              name="lastName"
              value="testlast"
            />
          </div>
          <div
            class="form-group"
          >
            <label>
              Email
            </label>
            <input
              class="form-control"
              name="email"
              value="test@test.net"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Confirm password to make changes:
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
          <button
            class="btn btn-primary btn-block mt-4"
          >
            Save Changes
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
